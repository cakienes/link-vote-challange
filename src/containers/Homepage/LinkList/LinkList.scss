.listOption {
    margin: 15px;
    outline: none;
}
