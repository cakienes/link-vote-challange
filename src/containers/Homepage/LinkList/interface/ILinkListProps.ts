import ILink from '../../../Interfaces/ILink';

export default interface ILinkListProps {
    links: ILink[];
}
