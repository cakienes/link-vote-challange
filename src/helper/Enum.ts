export enum VoteEnum {
    MOST_VOTED = 'MOST_VOTED',
    LESS_VOTED = 'LESS_VOTED',
}
