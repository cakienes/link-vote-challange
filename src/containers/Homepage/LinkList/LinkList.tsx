import React from 'react';
import { connect } from 'react-redux';
import { VoteEnum } from '../../../helper/Enum';
import ILink from '../../Interfaces/ILink';
import ILinkListProps from './interface/ILinkListProps';
import './LinkList.scss';
import LinkListTemplate from './LinkListTemplate/LinkListTemplate';
import PageBlock from './PageBlock/PageBlock';

const pageSize = 5;

interface ILinkListState {
    page: number;
    vote: string;
}
export class LinkList extends React.Component<ILinkListProps, ILinkListState> {
    constructor(props: ILinkListProps) {
        super(props);

        this.state = {
            page: 0,
            vote: '',
        };
    }
    render() {
        const { links } = this.props;
        const { page } = this.state;

        const totalSize = Math.floor((links.length - 1) / pageSize);
        return (
            <div>
                <div className="listOption">
                    <select onChange={this.onChangeSelect}>
                        <option value="">Order by</option>
                        <option value={VoteEnum.MOST_VOTED}>Most Voted (Z->A)</option>
                        <option value={VoteEnum.LESS_VOTED}>Less Voted (A->Z)</option>
                    </select>
                </div>
                <div className="listItems">{this.renderTemplate()}</div>
                <div className="linkList">
                    {totalSize >= 1 && (
                        <PageBlock currentPage={page} totalSize={totalSize} onClickPage={this.setPage} />
                    )}
                </div>
            </div>
        );
    }

    renderTemplate = (): React.ReactNode => {
        const { links } = this.props;
        const { page, vote } = this.state;
        if (links && links.length > 0) {
            let pageData: ILink[] = [];
            switch (vote) {
                case VoteEnum.LESS_VOTED:
                    pageData = links.sort(this.sortByLessVoted);
                    break;
                case VoteEnum.MOST_VOTED:
                    pageData = links.sort(this.sortByMostVoted);
                    break;
                default:
                    pageData = links.sort(this.sortByCreatedAt);
                    break;
            }

            pageData = pageData.slice(pageSize * page, pageSize * (page + 1));

            if (pageData.length === 0) {
                pageData = links.slice(pageSize * (page - 1), pageSize * page);
                this.setPage(page - 1);
            }
            return pageData.map((link: ILink) => <LinkListTemplate data={link} key={link.id} />);
        } else {
            return '';
        }
    };

    setPage = (page: number): void => {
        this.setState({
            page,
        });
    };

    onChangeSelect = (e: any): void => {
        this.setState({
            vote: e.target.value,
        });
    };

    sortByMostVoted = (x: ILink, y: ILink): number => {
        const value = y.point - x.point;
        if (value === 0) {
            return this.sortByUpdatedAt(x, y);
        }
        return value;
    };

    sortByLessVoted = (x: ILink, y: ILink): number => {
        const value = x.point - y.point;
        if (value === 0) {
            return this.sortByUpdatedAt(x, y);
        }
        return value;
    };

    sortByUpdatedAt = (x: ILink, y: ILink): number => {
        return new Date(y.updatedAt).getTime() - new Date(x.updatedAt).getTime();
    };

    sortByCreatedAt = (x: ILink, y: ILink): number => {
        return new Date(y.createdAt).getTime() - new Date(x.createdAt).getTime();
    };
}

const mapStateToProps = (state: any) => {
    return {
        links: state.link.links,
    };
};

export default connect(
    mapStateToProps,
    null,
)(LinkList);
