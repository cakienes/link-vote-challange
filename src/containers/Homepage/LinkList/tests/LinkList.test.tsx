import { mount, shallow } from 'enzyme';
import React from 'react';
import { VoteEnum } from '../../../../helper/Enum';
import ILink from '../../../Interfaces/ILink';
import ILinkListProps from '../interface/ILinkListProps';
import { LinkList } from '../LinkList';

describe('LinkList.tsx', () => {
    let linkListProps: ILinkListProps = {
        links: [],
    };

    beforeEach(() => {
        jest.spyOn(global.Date, 'now').mockImplementationOnce(() => new Date('2019-06-02T17:13:58.135Z').valueOf());
    });

    it('renders without crashing', async () => {
        expect(shallow(<LinkList {...linkListProps} />)).toMatchSnapshot();
    });

    it('setPage of LinkList', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance: any = wrapper.instance();

        expect(instance.state.page).toEqual(0);
        await instance.setPage(1);
        expect(instance.state.page).toEqual(1);
    });

    it('onChangeSelect of LinkList', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance: any = wrapper.instance();

        expect(instance.state.vote).toEqual('');

        const dumE = {
            target: {
                value: VoteEnum.MOST_VOTED,
            },
        };
        await instance.onChangeSelect(dumE);

        expect(instance.state.vote).toEqual(VoteEnum.MOST_VOTED);
    });

    it('sortByMostVoted of LinkList while points are not equal', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance: any = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();

        const dumLink1: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };
        const dumLink2: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 1,
            updatedAt: dumUpdatedDate,
        };
        expect(instance.sortByMostVoted(dumLink1, dumLink2)).toEqual(1);
    });

    it('sortByMostVoted of LinkList while points are equal', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance: any = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();

        const dumLink1: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };
        const dumLink2: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        instance.sortByUpdatedAt = jest.fn().mockReturnValue(1);
        await expect(instance.sortByMostVoted(dumLink1, dumLink2)).toEqual(1);
        expect(instance.sortByUpdatedAt).toHaveBeenCalledWith(dumLink1, dumLink2);
    });

    it('sortByLessVoted of LinkList while points are not equal', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance: any = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();

        const dumLink1: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 1,
            updatedAt: dumUpdatedDate,
        };
        const dumLink2: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };
        expect(instance.sortByLessVoted(dumLink1, dumLink2)).toEqual(1);
    });

    it('sortByLessVoted of LinkList while points are equal', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance: any = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();

        const dumLink1: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };
        const dumLink2: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        instance.sortByUpdatedAt = jest.fn().mockReturnValue(1);
        await expect(instance.sortByLessVoted(dumLink1, dumLink2)).toEqual(1);
        expect(instance.sortByUpdatedAt).toHaveBeenCalledWith(dumLink1, dumLink2);
    });

    it('sortByUpdatedAt of LinkList', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance: any = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();
        const dumLink1: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        const dumUpdatedDate2: Date = new Date(dumUpdatedDate.getDate() + 1);

        const dumLink2: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate2,
        };

        await expect(instance.sortByUpdatedAt(dumLink1, dumLink2)).toBeLessThan(0);
    });

    it('sortByCreatedAt of LinkList', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance: any = wrapper.instance();

        const dumCreatedDate = new Date();
        const dumUpdatedDate = new Date();
        const dumLink1: ILink = {
            createdAt: dumCreatedDate,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        const dumCreatedDate2: Date = new Date(dumCreatedDate.getDate() + 1);

        const dumLink2: ILink = {
            createdAt: dumCreatedDate2,
            id: 'id',
            link: 'link',
            name: 'name',
            point: 0,
            updatedAt: dumUpdatedDate,
        };

        await expect(instance.sortByCreatedAt(dumLink1, dumLink2)).toBeLessThan(0);
    });

    it('renderTemplate of LinkList while links is empty', async () => {
        const wrapper = mount(<LinkList {...linkListProps} />);
        const instance: any = wrapper.instance();

        await expect(instance.renderTemplate()).toEqual('');
    });
});
