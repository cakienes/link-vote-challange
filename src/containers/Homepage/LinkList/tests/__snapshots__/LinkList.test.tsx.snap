// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkList.tsx renders without crashing 1`] = `ShallowWrapper {}`;
